subprojects/core/src/integTest/groovy/org/gradle/api/invocation/GradleLifecycleAllprojectsIntegrationTest.groovy

```diff
@@ -0,0 +1,293 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.invocation
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import org.gradle.test.precondition.Requires
+import org.gradle.test.preconditions.IntegTestPreconditions
+
+@Requires(IntegTestPreconditions.NotIsolatedProjects)
```

Comment:
❓  Why?

Reply:
Because numerous `allprojects/suprojects` invocations?

Reply:
There are tests in this class that are compatible with IP, only the ones that access mutable state in those callbacks will fail. If we disable IP for this class, those tests won't be tested under IP.

For instance, `lifecycle.allprojects is executed before project.#api` is something we definitely want to validate both for vintage and IP.

Maybe it's the name of the test class that is too general? I imagine adding new tests related to `lifecycle.allprojects` there, and again it is strange if they would not be tested with IP, given that's the whole purpose.

How about we move the the tests that access mutable state to `GradleLifecycleAllprojectsEagerEvaluationIntegrationTest`? Then it makes a lot of sense to avoid running those tests with IP.

The tests that we'll keep in the general `GradleLifecycleAllprojectsIntegrationTest` should run both in vintage and IP.

Reply:
How the test `lifecycle.allprojects is executed before project.allprojects` can be successfully executed with IP enabled? Or I missing something... 🤔 

These tests indeed testing compatibility with eager APIs, so emphasize this in the name is a good point

Reply:
Sorry, I copied the wrong test name. Here are the current names of tests that do not use mutable state and make sense in the IP context:
- `lifecycle.allprojects is not executed again after project evaluation`
- `lifecycle.allprojects is executed only once for a project`
- `lifecycle.allprojects is executed before gradle.lifecycle.beforeProject`

Reply:
That's valid, I'll separate these tests

Reply:
@alllex Essentially, only `lifecycle.allprojects is executed before gradle.lifecycle.beforeProject` is IP compatible, the rest tests are touching mutable state(we expect lifecycle action execution, isn't it?). `project.name` is going throught `getProperty` path, so it's considering mutable state access. Although, `project.getName()` isn't

Reply:
> project.name is going throught getProperty path
Oh, we should fix that, I think.

The IP-version of this tries to be smarter by not issuing a violation when the property is found on the object itself. We should do a similar (the same?) thing here. Then `project.name` will not be considered mutable access.
https://github.com/gradle/gradle/blob/7f90032c8b91ef1fac29c4ad78c99da36d174618/platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/ProblemReportingCrossProjectModelAccess.kt#L209-L214

That is again why it is important to extract the "mutable access detecting project" logic into a separate class, and share it between both the new allprojects API and the cross-project access with IP.

Reply:
I think we are missing a test, where we access all things we consider immutable on another project (in both Groovy and Kotlin DSL), and then we check that the `lifecycle.allprojects` callback was not triggered.

This is also a test that makes sense in both vintage and IP.

See this test for IP:
https://github.com/gradle/gradle/blob/7f90032c8b91ef1fac29c4ad78c99da36d174618/platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/internal/cc/impl/isolated/IsolatedProjectsAccessFromGroovyDslIntegrationTest.groovy#L673